File: pkg/keys/recoverykey.go
package keys

// Recovery Key encoding — LONG-TERM COMPATIBILITY PROMISE.
//
// The RK is the only thing that can decrypt a user's backup (decisions.md), so
// its textual form must stay parseable by the offline decrypt CLI forever, and
// must be typable from a password manager.
//
// Format:
//
//	ocbk1-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX
//
//   - "ocbk1" is a versioned prefix; a future format bumps it to ocbk2.
//   - The payload is 160 bits of entropy + an 8-bit checksum = 168 bits,
//     encoded as 35 Crockford base32 characters in 7 groups of 5.
//   - Crockford base32 excludes I, L, O, and U, so it survives transcription;
//     decoding is case-insensitive and maps the look-alikes I/L->1 and O->0.
//   - The checksum catches typos before an expensive Argon2id derivation and
//     lets the UI say "that key is mistyped" rather than "wrong key".
//
// 160 bits of entropy is far beyond brute-force reach even though Argon2id also
// stretches it; the KDF exists to slow down attacks on the wrapped blob, not to
// compensate for weak entropy.
//
// The plaintext RK is generated client-side (Phase 8, WebCrypto) and never
// crosses the network. This file lives server-side only because the decrypt CLI
// links the same package.

import (
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"fmt"
	"strings"
)

const (
	// RKPrefix is the versioned Recovery Key prefix.
	RKPrefix = "ocbk1"
	// rkEntropyBytes is the raw entropy carried by an RK (160 bits).
	rkEntropyBytes = 20
	// rkGroupSize is the number of characters per dash-separated group.
	rkGroupSize = 5
	// crockford is the Crockford base32 alphabet (no I, L, O, U).
	crockford = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
)

// ErrBadRecoveryKey indicates a malformed, mistyped, or wrong-version RK.
var ErrBadRecoveryKey = errors.New("keys: invalid recovery key")

// GenerateRecoveryKey returns a fresh Recovery Key: the human-transportable
// string to show the user once, plus the raw secret bytes used for wrapping.
//
// The caller must show the string exactly once and never persist it server-side
// (decisions.md: no server-side RK escrow). Zeroize the returned secret when
// done.
func GenerateRecoveryKey() (display string, secret []byte, err error) {
	entropy := make([]byte, rkEntropyBytes)
	if _, err := rand.Read(entropy); err != nil {
		return "", nil, fmt.Errorf("keys: generate recovery key: %w", err)
	}
	return EncodeRecoveryKey(entropy), entropy, nil
}

// EncodeRecoveryKey renders raw entropy as the display form.
func EncodeRecoveryKey(entropy []byte) string {
	payload := make([]byte, 0, len(entropy)+1)
	payload = append(payload, entropy...)
	payload = append(payload, checksumByte(entropy))

	encoded := base32Encode(payload)
	var b strings.Builder
	b.WriteString(RKPrefix)
	for i := 0; i < len(encoded); i += rkGroupSize {
		end := min(i+rkGroupSize, len(encoded))
		b.WriteByte('-')
		b.WriteString(encoded[i:end])
	}
	return b.String()
}

// DecodeRecoveryKey parses a user-entered Recovery Key back into the raw secret
// used for wrapping. It is tolerant of case, spaces, and missing/extra dashes,
// and maps Crockford look-alike characters, because users retype these by hand.
//
// It never includes the input in its error, so a mistyped key cannot leak into
// logs (AGENTS.md: never log key material).
func DecodeRecoveryKey(s string) ([]byte, error) {
	cleaned := strings.TrimSpace(s)
	// Accept an optional, case-insensitive version prefix.
	lower := strings.ToLower(cleaned)
	switch {
	case strings.HasPrefix(lower, RKPrefix+"-"):
		cleaned = cleaned[len(RKPrefix)+1:]
	case strings.HasPrefix(lower, RKPrefix):
		cleaned = cleaned[len(RKPrefix):]
	default:
		// A different version prefix is an explicit, actionable failure.
		if i := strings.Index(lower, "-"); i > 0 && strings.HasPrefix(lower, "ocbk") {
			return nil, fmt.Errorf("%w: unsupported recovery key version", ErrBadRecoveryKey)
		}
	}

	var sb strings.Builder
	for _, r := range strings.ToUpper(cleaned) {
		switch r {
		case '-', ' ', '\t', '\n', '\r':
			continue
		case 'I', 'L':
			sb.WriteByte('1')
		case 'O':
			sb.WriteByte('0')
		default:
			if strings.ContainsRune(crockford, r) {
				sb.WriteRune(r)
				continue
			}
			return nil, fmt.Errorf("%w: illegal character", ErrBadRecoveryKey)
		}
	}

	payload, err := base32Decode(sb.String())
	if err != nil {
		return nil, err
	}
	if len(payload) != rkEntropyBytes+1 {
		return nil, fmt.Errorf("%w: wrong length", ErrBadRecoveryKey)
	}
	entropy := payload[:rkEntropyBytes]
	if payload[rkEntropyBytes] != checksumByte(entropy) {
		return nil, fmt.Errorf("%w: checksum mismatch (mistyped?)", ErrBadRecoveryKey)
	}
	return entropy, nil
}

// checksumByte is a truncated SHA-256 over the entropy, used only to detect
// transcription errors — it is not a security control.
func checksumByte(entropy []byte) byte {
	sum := sha256.Sum256(entropy)
	return sum[0]
}

// base32Encode encodes bytes MSB-first using the Crockford alphabet, without
// padding. Length is ceil(len(b)*8/5).
func base32Encode(b []byte) string {
	var out strings.Builder
	var acc uint32
	var bits uint
	for _, c := range b {
		acc = acc<<8 | uint32(c)
		bits += 8
		for bits >= 5 {
			bits -= 5
			out.WriteByte(crockford[(acc>>bits)&0x1f])
		}
	}
	if bits > 0 {
		out.WriteByte(crockford[(acc<<(5-bits))&0x1f])
	}
	return out.String()
}

// base32Decode reverses base32Encode. Input must already be upper-cased and
// contain only Crockford characters.
func base32Decode(s string) ([]byte, error) {
	out := make([]byte, 0, len(s)*5/8)
	var acc uint32
	var bits uint
	for i := 0; i < len(s); i++ {
		idx := strings.IndexByte(crockford, s[i])
		if idx < 0 {
			return nil, fmt.Errorf("%w: illegal character", ErrBadRecoveryKey)
		}
		acc = acc<<5 | uint32(idx)
		bits += 5
		if bits >= 8 {
			bits -= 8
			out = append(out, byte((acc>>bits)&0xff))
		}
	}
	// 21 payload bytes do not divide into 5-bit groups, so the final character
	// carries padding bits. base32Encode always writes them as zero; anything
	// else is a mistyped key. Without this check a typo confined to the padding
	// bits would decode to the same bytes and slip past the checksum.
	if bits > 0 && acc&((1<<bits)-1) != 0 {
		return nil, fmt.Errorf("%w: trailing bits (mistyped?)", ErrBadRecoveryKey)
	}
	return out, nil
}

File: web/src/crypto/recoverykey.ts
// Recovery Key encoding — the browser half of a LONG-TERM COMPATIBILITY
// PROMISE. Mirrors pkg/keys/recoverykey.go exactly; the offline decrypt CLI
// must keep parsing what this produces forever.
//
// Format:
//
//	ocbk1-XXXXX-XXXXX-XXXXX-XXXXX-XXXXX-XXXX
//
//   - "ocbk1" is a versioned prefix; a future format bumps it to ocbk2.
//   - The payload is 160 bits of entropy plus an 8-bit checksum = 168 bits,
//     encoded as 34 Crockford base32 characters: six groups of five and a
//     final group of four, 46 characters with the prefix and dashes.
//   - Crockford base32 excludes I, L, O and U, so the key survives handwriting;
//     decoding is case-insensitive and maps the look-alikes I/L -> 1, O -> 0.
//   - The checksum catches a typo before the expensive Argon2id derivation, so
//     the UI can say "mistyped" instead of "wrong key". It is an integrity aid,
//     not a security control.
//
// The wrapping secret is the 20 raw entropy bytes, never the display string.
import { sha256 } from '@noble/hashes/sha2.js'
import { randomBytes } from './bytes'

/** RK_PREFIX is the versioned Recovery Key prefix. */
export const RK_PREFIX = 'ocbk1'

/** RK_ENTROPY_BYTES is the raw entropy carried by a Recovery Key (160 bits). */
export const RK_ENTROPY_BYTES = 20

/** RK_GROUP_SIZE is the number of characters per dash-separated group. */
export const RK_GROUP_SIZE = 5

const CROCKFORD = '0123456789ABCDEFGHJKMNPQRSTVWXYZ'

/**
 * RecoveryKeyErrorReason says why a key was refused.
 *
 * These codes are shared with the Go implementation through
 * pkg/keys/testdata/vectors.json. Two implementations will never phrase an
 * error identically, but they must agree on *why* a key failed: "mistyped" and
 * "wrong key" lead a user to opposite actions.
 */
export type RecoveryKeyErrorReason =
  | 'length'
  | 'version'
  | 'charset'
  | 'checksum'
  | 'padding'

/**
 * RecoveryKeyError is a malformed, mistyped, or wrong-version Recovery Key.
 *
 * It never includes the input, so a mistyped key cannot reach a log or an
 * error-reporting service.
 */
export class RecoveryKeyError extends Error {
  readonly reason: RecoveryKeyErrorReason

  constructor(reason: RecoveryKeyErrorReason, message: string) {
    super(message)
    this.name = 'RecoveryKeyError'
    this.reason = reason
  }
}

/** RecoveryKey is a freshly generated key: what to show, and what to wrap with. */
export interface RecoveryKey {
  /** display is the string to show the user exactly once. */
  display: string
  /** secret is the raw entropy used for wrapping. Zeroize it when done. */
  secret: Uint8Array
}

/**
 * generateRecoveryKey draws a fresh Recovery Key.
 *
 * The caller must show `display` once and never send it anywhere: there is no
 * server-side escrow, by design (decisions.md).
 */
export function generateRecoveryKey(): RecoveryKey {
  const secret = randomBytes(RK_ENTROPY_BYTES)
  return { display: encodeRecoveryKey(secret), secret }
}

/** encodeRecoveryKey renders raw entropy as the display form. */
export function encodeRecoveryKey(entropy: Uint8Array): string {
  if (entropy.length !== RK_ENTROPY_BYTES) {
    throw new RecoveryKeyError('length', `recovery key entropy must be ${RK_ENTROPY_BYTES} bytes`)
  }

  const payload = new Uint8Array(entropy.length + 1)
  payload.set(entropy, 0)
  payload[entropy.length] = checksumByte(entropy)

  const encoded = base32Encode(payload)
  let out = RK_PREFIX
  for (let i = 0; i < encoded.length; i += RK_GROUP_SIZE) {
    out += '-' + encoded.slice(i, i + RK_GROUP_SIZE)
  }
  return out
}

/**
 * decodeRecoveryKey parses a user-entered Recovery Key back into the raw
 * secret. It tolerates case, spaces and missing or extra dashes, and maps the
 * Crockford look-alikes, because users retype these by hand from paper.
 */
export function decodeRecoveryKey(input: string): Uint8Array {
  let cleaned = input.trim()
  const lower = cleaned.toLowerCase()

  if (lower.startsWith(RK_PREFIX + '-')) {
    cleaned = cleaned.slice(RK_PREFIX.length + 1)
  } else if (lower.startsWith(RK_PREFIX)) {
    cleaned = cleaned.slice(RK_PREFIX.length)
  } else if (lower.startsWith('ocbk') && lower.indexOf('-') > 0) {
    // A different version prefix is an explicit, actionable failure: the key is
    // fine, this client is too old to read it.
    throw new RecoveryKeyError('version', 'unsupported recovery key version')
  }

  let symbols = ''
  for (const character of cleaned.toUpperCase()) {
    switch (character) {
      case '-':
      case ' ':
      case '\t':
      case '\n':
      case '\r':
        continue
      case 'I':
      case 'L':
        symbols += '1'
        break
      case 'O':
        symbols += '0'
        break
      default:
        if (!CROCKFORD.includes(character)) {
          throw new RecoveryKeyError('charset', 'recovery key contains an illegal character')
        }
        symbols += character
    }
  }

  const payload = base32Decode(symbols)
  if (payload.length !== RK_ENTROPY_BYTES + 1) {
    throw new RecoveryKeyError('length', 'recovery key has the wrong length')
  }

  const entropy = payload.subarray(0, RK_ENTROPY_BYTES)
  if (payload[RK_ENTROPY_BYTES] !== checksumByte(entropy)) {
    throw new RecoveryKeyError('checksum', 'recovery key checksum mismatch (mistyped?)')
  }
  return Uint8Array.from(entropy)
}

/** checksumByte is a truncated SHA-256 over the entropy: a typo detector. */
function checksumByte(entropy: Uint8Array): number {
  return sha256(entropy)[0] as number
}

/** base32Encode encodes MSB-first with the Crockford alphabet, unpadded. */
function base32Encode(bytes: Uint8Array): string {
  let out = ''
  let accumulator = 0
  let bits = 0
  for (const byte of bytes) {
    accumulator = (accumulator << 8) | byte
    bits += 8
    while (bits >= 5) {
      bits -= 5
      out += CROCKFORD[(accumulator >>> bits) & 0x1f]
    }
  }
  if (bits > 0) {
    out += CROCKFORD[(accumulator << (5 - bits)) & 0x1f]
  }
  return out
}

/** base32Decode reverses base32Encode. Input is already upper-cased and filtered. */
function base32Decode(symbols: string): Uint8Array {
  const out: number[] = []
  let accumulator = 0
  let bits = 0
  for (const symbol of symbols) {
    const index = CROCKFORD.indexOf(symbol)
    if (index < 0) {
      throw new RecoveryKeyError('charset', 'recovery key contains an illegal character')
    }
    accumulator = (accumulator << 5) | index
    bits += 5
    if (bits >= 8) {
      bits -= 8
      out.push((accumulator >>> bits) & 0xff)
    }
  }
  // 21 payload bytes do not divide into 5-bit groups, so the last character
  // carries padding bits. They are always written as zero; anything else is a
  // typo confined to those bits, which would otherwise decode to the same
  // entropy and slip past the checksum.
  if (bits > 0 && (accumulator & ((1 << bits) - 1)) !== 0) {
    throw new RecoveryKeyError('padding', 'recovery key has non-zero trailing bits (mistyped?)')
  }
  return Uint8Array.from(out)
}
